# Basic `pandas`

In this notebook we will learn about `pandas` a vital data management package in python.

By the end of this notebook you will know about:
- `pandas` `Series` and `DataFrame` objects,
- Indexing `pandas` objects,
- Reading files with `pandas`,
- Saving files with `pandas` and 
- Some helpful `pandas` functionality.

## `pandas`

`pandas` is one of the most popular data handling packages in `python`. We will cover the minimum you need to know about the package for the boot camp in this notebook.

Let's start by importing the package. Just like in the previous notebook this should check whether or not you have `pandas` installed on your machine. If you are using Anaconda, <a href="https://www.anaconda.com/">https://www.anaconda.com/</a>, `pandas` should be installed already. If not check out the documentation for installation instructions <a href="https://pandas.pydata.org/docs/getting_started/install.html">https://pandas.pydata.org/docs/getting_started/install.html</a>.

In [2]:
## it is standard to import pandas as pd
import pandas as pd

In [3]:
## let's check what version of numpy you have
## when I wrote this I had version 1.5.3
## yours may be different

print(pd.__version__)

2.1.4


If you had a version of `pandas` installed, both of those code chunks should have executed without error. If not, you will need to install it onto your machine because we will be using it heavily in the boot camp. If you are unsure how to install a python package in general check our python package installation guide, <a href="https://www.erdosinstitute.org/data-science">https://www.erdosinstitute.org/data-science</a>.

##### Be sure you can run both of the above code chunks before continuing with this notebook, again it should be fine if your package version is slightly different than mine.

### `Series` and `DataFrame`s

`pandas` has two main data structures: 
1. `Series` objects, <a href="https://pandas.pydata.org/docs/reference/api/pandas.Series.html">https://pandas.pydata.org/docs/reference/api/pandas.Series.html</a> and 
2. `DataFrame` objects, <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html</a>. 

Let's explore them below.

In [4]:
## We can turn a list into a series
## with pd.Series()
list1 = [0,1,2,3]
print("Below is a list.")
print(list1, type(list1))
print()
print("Below is a pandas Series object.")
print(pd.Series(list1), type(pd.Series(list1)))

Below is a list.
[0, 1, 2, 3] <class 'list'>

Below is a pandas Series object.
0    0
1    1
2    2
3    3
dtype: int64 <class 'pandas.core.series.Series'>


The second thing we printed was a `Series` object. Note the two columns of numbers. The first column is the index of the object, the second column contains the values of the object. We can access those two separately like below.

In [5]:
series1 = pd.Series(list1)

In [6]:
## The index
series1.index

RangeIndex(start=0, stop=4, step=1)

In [7]:
## The values
series1.values

array([0, 1, 2, 3], dtype=int64)

In [9]:
## You practice
## Take the list labeled a and 
## turn it into a Series named b
a = [5,2,3,6,'a','b','e',True,False]

b = pd.Series(a)
b

0        5
1        2
2        3
3        6
4        a
5        b
6        e
7     True
8    False
dtype: object

In [10]:
## We can get select entries from a Series 
## using the index

b[5]

'b'

In [11]:
## Note that we don't have to have our index be a list of numbers
c = pd.Series([1,2,3,4], index=['a','b','c','d'])

print(c)

a    1
b    2
c    3
d    4
dtype: int64


In [17]:
## and we can access by index in the same way
c['a']

1

Now let's check out a `DataFrame`. `DataFrame`s are essentially a collection of `Series` with a common index. We can also think of them as a table of values with column and row labels.

In [18]:
## We can make a DataFrame using a dictionary
## the dictionary keys are the column labels
## the dictionary values are columns
df = pd.DataFrame({'one':[3,4,5,2,4,5], 
                    'two':['a','b','e','h','l','p'],
                    'third column':[7,7,7,7,7,7]})

## NOTE that this is not the only way to make 
## a dataframe!

print(df)

   one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7


This is a `DataFrame`, the unlabeled column is the index, the labeled columns are `Series` objects themselves. We can access them in the following way.

In [19]:
## If your column's name doesn't violate a couple of format rules
## you can use
## df.column_name
print(df.one) 
print()
print("Note that each column is a Series object.")
print(type(df.one))

0    3
1    4
2    5
3    2
4    4
5    5
Name: one, dtype: int64

Note that each column is a Series object.
<class 'pandas.core.series.Series'>


In [20]:
## However if our column name has spaces or certain characters in it
## (like . , ! ? "" and so on)
## we can't use .column_name

df.third column

SyntaxError: invalid syntax (3086096821.py, line 5)

In [21]:
## So we have to use df["column name"] instead
# or df['column_name']

df["third column"]

0    7
1    7
2    7
3    7
4    7
5    7
Name: third column, dtype: int64

In [22]:
## Just like with a Series we can use .index
## to get the index
df.index

RangeIndex(start=0, stop=6, step=1)

In [23]:
## .values returns a 2-D numpy ndarray with our columns
df.values

array([[3, 'a', 7],
       [4, 'b', 7],
       [5, 'e', 7],
       [2, 'h', 7],
       [4, 'l', 7],
       [5, 'p', 7]], dtype=object)

In [26]:
## You code
## Make a data frame, call it my_df 
## Make the first column labeled 'first' from a
## Make the second column labeled 'second' from b
## see what happens when you add , index=range(10,10+len(a)) 
## after the dictionary
a = [4,5,3,4,5,6,0]
b = ['a','c','d','g','l','m','p']

my_df = pd.DataFrame({"first": a, "second": b}, index = range(10,10+len(a)))


In [27]:
my_df

first second
10      4      a
11      5      c
12      3      d
13      4      g
14      5      l
15      6      m
16      0      p

#### Locating `DataFrame` Entries With the Index

Locating entries in a `DataFrame` is slightly more complicated than what we had to do for `Series` objects.

In [29]:
## You code
## try to index the dataframe df like you would
## a list or Series to find the entry corresponding
## to index 1
df[1]

## What happens?

KeyError: 1

In order to get particular `DataFrame` entries you have to use one of two methods, `.loc[]` or `.iloc[]`.

In [30]:
df

one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7

In [31]:
## for example we can get the 1 row of df like so with loc

df.loc[1]

one             4
two             b
third column    7
Name: 1, dtype: object

In [32]:
## or like so with iloc

df.iloc[1]

one             4
two             b
third column    7
Name: 1, dtype: object

The difference between `.loc` and `.iloc` is that `.iloc` is restricted to integer based indexing, while `.loc` is more versatile.

This means that if your index is a list of strings you cannot use `.iloc` with just the index's normal value. For example try to run the following.

In [33]:
## Recall what c is

c

a    1
b    2
c    3
d    4
dtype: int64

In [34]:
## now try to get the ath entry using .iloc

c.iloc["a"]

TypeError: Cannot index by location index with a non-integer key

In [35]:
## instead we'd have to notice that the "a" entry is
## the 0 entry if the index were a normal integer based index
c.iloc[0]

1

In [36]:
## However, we can just insert "a" into .loc
c.loc["a"]

1

Another difference is that `.loc` can be used to perform boolean indexing (eg. find me all the rows where the first column is between $2$ and $4$)

In [39]:
df

one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7

In [40]:
## give me all the rows where the "one" column is between 2 and 4
##multiple logical conditions must be within parentheses
df.loc[(df.one > 2) & (df.one < 4)]

one two  third column
0    3   a             7

In [41]:
## we can even go one step further and find a specific column
df.loc[(df.one > 2) & (df.one < 4), "third column"]

0    7
Name: third column, dtype: int64

In [42]:
## or a subset of many columns
## in this case you enter the subset as a list of column name strings
df.loc[(df.one > 2) & (df.one < 4), ["one","third column"]]

one  third column
0    3             7

In [43]:
## order of columns can even be different from dataframe
df.loc[(df.one < 4) & (df.one > 2), ['third column','two']]

third column two
0             7   a

In [44]:
#also can provide multiple columns in df[]

df[["third column", "two"]]

third column two
0             7   a
1             7   b
2             7   e
3             7   h
4             7   l
5             7   p

### Reading in a `.csv` file with `pandas`

You can also read in common data file types with `pandas`. Let's load the following.

In [45]:
## You can read in a csv with
## pd.read_csv("filename.csv")
## https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

## We'll work with the following dataframe in the next section
jr_shots = pd.read_csv("../data/JR_Smith_Shots_2015_16.csv")

In [46]:
jr_shots

LOC_X  LOC_Y  SHOT_MADE_FLAG
0     -106    244               0
1      -96     97               0
2       30     23               0
3     -204     -1               0
4      -76    237               0
..     ...    ...             ...
844    145    -15               0
845   -241     67               0
846    164    195               0
847      0      1               1
848    -89    288               0

[849 rows x 3 columns]

In [47]:
## You code
## read in the file "beer.csv" using pandas
## call it beer_df

beer_df = pd.read_csv("../data/beer.csv")

## you'll practice with this dataframe in the next section

In [48]:
beer_df

IBU   ABV  Rating Beer_Type
0     45   4.2   3.792     Stout
1     60   8.3   4.145     Stout
2     25   6.0   3.951     Stout
3     31  11.0   4.062     Stout
4     75   9.0   4.018     Stout
..   ...   ...     ...       ...
342   68   8.2   3.922       IPA
343   45   6.8   4.000       IPA
344  101   9.2   3.778       IPA
345   70   6.8   4.015       IPA
346   80   6.9   3.554       IPA

[347 rows x 4 columns]

There are other useful `read` functions like `read_table` and `read_json`, you will see those later.

### Helpful `pandas` functions

`pandas` offers some really nice built in functions to help you explore any data set you're dealing with. Let's explore them below.

In [51]:
## df.head(n) let's you inspect the first n rows of the dataframe
## n defaults to 5
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html
jr_shots.head()

LOC_X  LOC_Y  SHOT_MADE_FLAG
0   -106    244               0
1    -96     97               0
2     30     23               0
3   -204     -1               0
4    -76    237               0

In [54]:
## You code 
## investigate what .tail(n) does 
## using the beer_df
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html
## Hint: n must be an integer

beer_df.tail()

IBU  ABV  Rating Beer_Type
342   68  8.2   3.922       IPA
343   45  6.8   4.000       IPA
344  101  9.2   3.778       IPA
345   70  6.8   4.015       IPA
346   80  6.9   3.554       IPA

In [56]:
## You code
## investigate what .sample(n) does
## using the beer_df
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html
## Hint: n must be an integer

beer_df.sample(3)

IBU  ABV  Rating Beer_Type
14   28  5.0   4.075     Stout
92   30  8.0   3.639     Stout
97   30  7.4   3.810     Stout

In [67]:
## You code
## what happens when you add the argument
## random_state=440 in .sample()?
## rerun the code chunk multiple times with or without the random_state
## try different numbers for the random_state

beer_df.sample(5, random_state=441)

IBU   ABV  Rating Beer_Type
81    25  10.5   4.310     Stout
256   57   5.7   3.572       IPA
311  100   9.0   4.265       IPA
109   50   8.0   3.892     Stout
259   95   7.0   3.865       IPA

In [68]:
## df.info() tells you useful information about your
## dataframe, like the column names,
## the number of rows,
## the number of non-empty entries for each column,
## the data type of each column
## and how much memory the dataframe uses
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html
jr_shots.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 849 entries, 0 to 848
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   LOC_X           849 non-null    int64
 1   LOC_Y           849 non-null    int64
 2   SHOT_MADE_FLAG  849 non-null    int64
dtypes: int64(3)
memory usage: 20.0 KB


In [69]:
## df.describe()
## provides the 5 stat breakdown of each numeric column
## that is any column that consists of integers or floats
## will have its, minimum, maximum, first quartile, median,
## third quartile, and mean provided
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html
jr_shots.describe()

LOC_X       LOC_Y  SHOT_MADE_FLAG
count  849.000000  849.000000      849.000000
mean    -6.204947  112.106007        0.414605
std    157.201943   91.254429        0.492944
min   -250.000000  -46.000000        0.000000
25%   -158.000000   21.000000        0.000000
50%     -2.000000  110.000000        0.000000
75%    133.000000  193.000000        1.000000
max    245.000000  693.000000        1.000000

In [88]:
## You code
## Try out .mean() and .max() and .min()
## on beer_df
## Compare the results to .describe()'s output

## .mean() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
## .max() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.max.html
## .min() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.min.html
print("Mean:")
print(beer_df.mean(numeric_only=True))

print("Max:")
print(beer_df.max())

print("Min:")
print(beer_df.min())

beer_df.describe()

Mean:
IBU       57.579251
ABV        7.639769
Rating     3.853980
dtype: float64
Max:
IBU            250
ABV           18.0
Rating       4.541
Beer_Type    Stout
dtype: object
Min:
IBU              6
ABV            3.8
Rating       3.022
Beer_Type      IPA
dtype: object


IBU         ABV      Rating
count  347.000000  347.000000  347.000000
mean    57.579251    7.639769    3.853980
std     24.100858    2.216800    0.228915
min      6.000000    3.800000    3.022000
25%     40.000000    6.200000    3.700500
50%     60.000000    7.000000    3.823000
75%     70.000000    8.850000    3.982500
max    250.000000   18.000000    4.541000

In [89]:
## What about variables that aren't numeric? Like
## classes
## You can use df.value_counts() for those
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.value_counts.html
jr_shots.SHOT_MADE_FLAG.value_counts()

SHOT_MADE_FLAG
0    497
1    352
Name: count, dtype: int64

The above tells us that 497 observations have a `SHOT_MADE_FLAG` of 0 and 352 have a `SHOT_MADE_FLAG` of 1.

In [100]:
## You code
## How many of each "Beer_Type" are there in beer_df?

print(beer_df.Beer_Type.value_counts())
print()
## What happens when you input the argument normalize=True into value_counts()?

print(beer_df.Beer_Type.value_counts(normalize=True))
print()

Beer_Type
IPA      195
Stout    152
Name: count, dtype: int64

Beer_Type
IPA      0.56196
Stout    0.43804
Name: proportion, dtype: float64



One task that we will do quite often in the boot camp is split one data set into two non-overlapping data sets. Let's see how we can do that here.

In [107]:
## Let's first randomly sample 100 observations from jr_shots
## Note the .copy()
## this creates a deep copy of the DataFrame 
jr_1 = jr_shots.sample(100, random_state=54888).copy()
jr_1
jr_1.index

Index([302, 474, 647, 109, 376, 673, 732, 741, 787, 235,  85, 700, 645, 629,
       269, 468,  79, 412,  58, 746, 144,  77, 168, 390,  57, 717, 724,   8,
       175, 557, 473, 286, 157, 828, 332, 671, 362, 773, 220, 108, 748, 400,
       822, 500, 541, 836, 545, 152, 483, 113, 389, 753, 507, 299, 843, 173,
       265,  96, 772, 719,  81, 674, 502, 761, 310, 198, 740, 306,  31, 771,
       664, 457, 444,  21, 258, 832, 147,  38, 383, 680, 603, 762,  97, 847,
       588, 574, 565, 816, 730, 492, 515, 320, 488, 536, 614, 149, 766, 792,
       336, 342],
      dtype='int64')

In [109]:
## To create jr_2 we'll just drop the indices of jr_1 from
## jr_shots using .drop()
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html

## Again we include .copy() to make a deep copy of the dataframe
jr_2 = jr_shots.drop(jr_1.index).copy()
jr_2

LOC_X  LOC_Y  SHOT_MADE_FLAG
0     -106    244               0
1      -96     97               0
2       30     23               0
3     -204     -1               0
4      -76    237               0
..     ...    ...             ...
842    171    178               1
844    145    -15               0
845   -241     67               0
846    164    195               0
848    -89    288               0

[749 rows x 3 columns]

In [110]:
jr_1

LOC_X  LOC_Y  SHOT_MADE_FLAG
302    227     -1               0
474    -81    174               0
647   -241    111               0
109     15     65               1
376     60     21               1
..     ...    ...             ...
149   -165    198               0
766    128    188               0
792    140    208               0
336     25      8               0
342    -40    180               1

[100 rows x 3 columns]

In [111]:
jr_2

LOC_X  LOC_Y  SHOT_MADE_FLAG
0     -106    244               0
1      -96     97               0
2       30     23               0
3     -204     -1               0
4      -76    237               0
..     ...    ...             ...
842    171    178               1
844    145    -15               0
845   -241     67               0
846    164    195               0
848    -89    288               0

[749 rows x 3 columns]

In [112]:
## Let's check that it worked like we thought it would
len(jr_1) + len(jr_2) == len(jr_shots)

True

In [115]:
## You code
## split beer_df into beer_1 and beer_2
## beer_1 should be a random sample of 150 observations
## beer_2 should be all rows not in beer_1
beer_1 = beer_df.sample(150).copy()

print(beer_1)
print()

beer_2 = beer_df.drop(beer_1.index).copy()

print(beer_2)
print()


     IBU   ABV  Rating Beer_Type
155   70  7.00   3.986       IPA
266   42  7.50   3.787       IPA
34    37  5.20   3.770     Stout
238   86  9.40   3.972       IPA
199   60  7.10   3.598       IPA
..   ...   ...     ...       ...
328   70  7.00   3.785       IPA
254   66  7.00   3.588       IPA
274   73  7.25   3.902       IPA
93    65  8.70   3.958     Stout
236   55  6.70   3.616       IPA

[150 rows x 4 columns]

     IBU  ABV  Rating Beer_Type
0     45  4.2   3.792     Stout
1     60  8.3   4.145     Stout
2     25  6.0   3.951     Stout
8     39  5.9   3.997     Stout
9     15  5.7   3.728     Stout
..   ...  ...     ...       ...
341   70  7.2   3.549       IPA
342   68  8.2   3.922       IPA
344  101  9.2   3.778       IPA
345   70  6.8   4.015       IPA
346   80  6.9   3.554       IPA

[197 rows x 4 columns]



In [116]:
len(beer_1) + len(beer_2) == len(beer_df)



True

### Saving a `DataFrame` to File

Just like we easily read in a data file we can use `pandas` to quickly save data to a file.

This is done with the function `df.to_csv()`, <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html</a>.

In [117]:
## Let's first make a data frame
df = pd.DataFrame({'a':[1,2,3],
                   'b':[2,4,6],
                   'c':[17,34,51]})

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

In [119]:
## Now we call df.to_csv(file_name)
df.to_csv("../data/our_first_dataframe.csv")

Go ahead and check your repository, you should now see `our_first_dataframe.csv` in there.

In [121]:
## You code
## read in "our_first_dataframe.csv" using read_csv
## then look at the df

df = pd.read_csv("../data/our_first_dataframe.csv")
df

## is anything off?

Unnamed: 0  a  b   c
0           0  1  2  17
1           1  2  4  34
2           2  3  6  51

When you ran `to_csv()` the default is to record the index as well. That is why your `DataFrame` has an unlabeled column when you read it back in using `read_csv`. If we wanted to avoid writing the index to file you include the argument `index=False`. 

In [122]:
## Make the DataFrame again
df = pd.DataFrame({'a':[1,2,3],
                   'b':[2,4,6],
                   'c':[17,34,51]})

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

In [123]:
## write it to file without the index
df.to_csv("../data/our_first_dataframe.csv",index=False)

In [124]:
## You code
## read in "our_first_dataframe.csv" using read_csv
## then look at the df again

df= pd.read_csv("../data/our_first_dataframe.csv")

df



a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

That's all you will need to know from `pandas` to have a firm footing for our boot camp. We will learn more stuff as we go along, but when that time comes you should be ready!

--------------------------

This notebook was written for the Erd&#337;s Institute C&#337;de Data Science Boot Camp by Matthew Osborne, Ph. D., 2023.

Any potential redistributors must seek and receive permission from Matthew Tyler Osborne, Ph.D. prior to redistribution. Redistribution of the material contained in this repository is conditional on acknowledgement of Matthew Tyler Osborne, Ph.D.'s original authorship and sponsorship of the Erdős Institute as subject to the license (see License.md)